### PR TITLE
Less measure inference

### PR DESCRIPTION
**Pull request by jad-hamza**

### CODE REVIEW

**Comment by romac** [comment]
I had tried that, but it seems to me that we actually need to pass functions which already have a measure through the whole phase as well, as we sometimes need the `DecreasesProcessor` to rewrite measures of inter-dependent functions to give them the same shape.

**Comment by jad-hamza** [comment]
What if we remove the measures for these inter-dependent functions?

**Comment by romac** [comment]
Well, then we wouldn't be able to provide measures for functions for which we cannot infer them.

But perhaps we could force the programmer to write measures of the same shapes for mutually recursive groups of functions?

**Comment by jad-hamza** [comment]
> But perhaps we could force the programmer to write measures of the same shapes for mutually recursive groups of functions?

Yes this makes sense. I think either the programmer either gives the same measure type for all mutually recursive functions, or no measure at all for these functions.

**Comment by jad-hamza** [comment]
> Well, then we wouldn't be able to provide measures for functions for which we cannot infer them.

What do you mean?

**Comment by romac** [comment]
> > But perhaps we could force the programmer to write measures of the same shapes for mutually recursive groups of functions?
> 
> Yes this makes sense. I think either the programmer either gives the same measure type for all mutually recursive functions, or no measure at all for these functions.

Alright, then I'll see if I can get to it next week, as I am in the middle of cutting a new release right now.

**Comment by romac** [comment]
> > Well, then we wouldn't be able to provide measures for functions for which we cannot infer them.
> 
> What do you mean?

Perhaps I misunderstood what you meant. I was saying that if we automatically remove measures on functions which already have some, then we cannot override the measure inference mechanism.

**Comment by jad-hamza** [comment]
Ah, I meant we remove them at the user-level (for the ones that can be inferred)

**Comment by romac** [comment]
> Ah, I meant we remove them at the user-level (for the ones that can be inferred)

Sure, that makes more sense :)

**Comment by jad-hamza** [comment]
This is going through now, I added a few improvements for error handling (I didn't test them a lot). Now the phase `MeasureInference` complains if a function has a measure but a mutually recursive function doesn't. The typechecker doesn't crash when a measure is missing, but adds a dummy (false) VC when a measure is missing.

File: core/src/main/scala/stainless/termination/MeasureInference.scala [hunk 1]
```diff
@@ -54,7 +59,26 @@ trait MeasureInference
...
     }
 
-    def needsMeasure(fd: FunDef): Boolean = symbols.isRecursive(fd.id)
+    def needsMeasure(fd: FunDef): Boolean = {
+      if (symbols.isRecursive(fd.id) && fd.measure(symbols).isEmpty) {
+        symbols.dependencies(fd.id).find(id =>
+          symbols.lookupFunction(id).exists(fd2 =>
+            symbols.dependencies(fd2.id).contains(fd.id) &&
+            !fd2.measure(symbols).isEmpty
+          )
```
Comment (romac) [inline core/src/main/scala/stainless/termination/MeasureInference.scala:68]: This block could perhaps be given a name to improve readability?

Comment (jad-hamza) [reply core/src/main/scala/stainless/termination/MeasureInference.scala]: Thanks!

**Pull request merged**